[NCCL] Add Environment Variable to guard Async Error Handling feature

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44163 [NCCL] Add Environment Variable to guard Async Error Handling feature**
* #41054 [NCCL] Destructor Blocks on WorkNCCL Completion
* #41053 [NCCL] WorkNCCL Helper Functions
* #41052 [NCCL] Abort Errored and Timed Out NCCL Communicators from Watchdog Thread
* #41051 [NCCL] Use cudaEventQuery to Poll for GPU operation errors
* #41050 [NCCL] Timeout Loop Thread for Async Error Handling

In this PR, we introduce a new environment variable
(NCCL_ASYNC_ERROR_HANDLING), which guards the asynchronous error handling
feature. We intend to eventually turn this feature on by default for all users,
but this is a temporary solution so the change in behavior from hanging to
crashing is not the default for users all of a sudden.

Differential Revision: [D23517895](https://our.internmc.facebook.com/intern/diff/D23517895/)

osalpekar committed:
[NCCL] Add Environment Variable to guard Async Error Handling feature

In this PR, we introduce a new environment variable
(NCCL_ASYNC_ERROR_HANDLING), which guards the asynchronous error handling
feature. We intend to eventually turn this feature on by default for all users,
but this is a temporary solution so the change in behavior from hanging to
crashing is not the default for users all of a sudden.

Differential Revision: [D23517895](https://our.internmc.facebook.com/intern/diff/D23517895/)

[ghstack-poisoned]

jiayisuse commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -404,6 +404,20 @@ void ProcessGroupNCCL::parseNcclBlockingWait() {
   }
 }
 
+void ProcessGroupNCCL::parseNcclAsyncErrorHandling() {
```
This function looks good. Wondering if we can reuse parseNcclBlockingWait() somehow

osalpekar replied:
I think it should be doable, I've made an issue to follow up on this: #44205 (mainly because we might be able to make this helper function more broadly available in the PyTorch codebase)

jiayisuse commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -428,6 +442,13 @@ ProcessGroupNCCL::ProcessGroupNCCL(
         std::string(NCCL_BLOCKING_WAIT));
   }
+  try {
+    parseNcclAsyncErrorHandling();
```
We could add log if async error handling is enabled

osalpekar committed:
Update on "[NCCL] Add Environment Variable to guard Async Error Handling feature"


In this PR, we introduce a new environment variable
(NCCL_ASYNC_ERROR_HANDLING), which guards the asynchronous error handling
feature. We intend to eventually turn this feature on by default for all users,
but this is a temporary solution so the change in behavior from hanging to
crashing is not the default for users all of a sudden.

Differential Revision: [D23517895](https://our.internmc.facebook.com/intern/diff/D23517895/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Add Environment Variable to guard Async Error Handling feature"


In this PR, we introduce a new environment variable
(NCCL_ASYNC_ERROR_HANDLING), which guards the asynchronous error handling
feature. We intend to eventually turn this feature on by default for all users,
but this is a temporary solution so the change in behavior from hanging to
crashing is not the default for users all of a sudden.

Differential Revision: [D23517895](https://our.internmc.facebook.com/intern/diff/D23517895/)

[ghstack-poisoned]
